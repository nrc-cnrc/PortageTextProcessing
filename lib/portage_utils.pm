#!/usr/bin/perl -w

# $Id$

# @file portage_utils.pm
# @brief Library to transparently use compress formats.
#
# @author Samuel Larkin
#
# Technologies langagieres interactives / Interactive Language Technologies
# Inst. de technologie de l'information / Institute for Information Technology
# Conseil national de recherches Canada / National Research Council Canada
# Copyright 2009, Sa Majeste la Reine du Chef du Canada /
# Copyright 2009, Her Majesty in Right of Canada

package portage_utils;
require Exporter;
@ISA = qw(Exporter);
@EXPORT = qw(zopen);  # symbols to export on request

use strict;
use warnings;



# How to detect that a file was gzipped.
my $isZip = qr/\.(gz|z|Z)\s*$/;

# How to detect that a file was gzipped.
my $isBzip2 = qr/\.(bz2|bzip2|bz)\s*$/;

# How to detect that a file was gzipped.
my $isLzma = qr/\.lzma\s*$/;

our $DEBUG;


# ================= zopen ====================#

=head1 SUB

B< =============================================
 ====== zopen                             ======
 =============================================>

=over 4

=item B<DESCRIPTION>

 Portage's perl magicstream.  Tries to deduce if the stream is input or output,
 if it is compressed (.gz,z,Z,bz2,bzip2,bz,lzma), piped or just a normal file,
 and does the proper thing to open the stream.

=item B<SYNOPSIS>

 portage_utils::zopen($STREAM, $stream_name);
 portage_utils::zopen(*STREAM, $stream_name);
 portage_utils::zopen(*STREAM, "plain-text-input-file");
 portage_utils::zopen(*STREAM, "< plain-text-input-file");
 portage_utils::zopen(*STREAM, "compressed-input-file.gz");
 portage_utils::zopen(*STREAM, "< compressed-input-file.gz");
 portage_utils::zopen(*STREAM, "input pipe |");
 portage_utils::zopen(*STREAM, "> plain-text-output-file");
 portage_utils::zopen(*STREAM, "> compressed-output-file.gz");
 portage_utils::zopen(*STREAM, "| output pipe ");

 @PARAM $STREAM
 @PARAM $stream_name

 @RETURN Returns 0 if the stream was opened.

=item B<SEE ALSO>

 zout()

=back

=cut

sub zopen($$) {
   # if no argument is sent in
   if(not @_) {
      die "\n!!! ERROR: portage_utils::zopen requires input arguments! \taborting...\n";
   }
   # if we are called as a module then skip the our reference name
   elsif($_[0] and $_[0] =~ /^portage_utils/) {
      shift;
      # if no argument is sent in
      if(not @_) {
         die "\n!!! ERROR: portage_utils::zopen requires input arguments! \taborting...\n";
      }
   }

   my($STREAM, $stream_name) = @_;

   if ($stream_name =~ /^\s*(>|\|)/) {
      return zout($STREAM, $stream_name);
   }
   else {
      return zin($STREAM, $stream_name);
   }
}

# ================= zin ====================#

=head1 SUB

B< =============================================
 ====== zin                             ======
 =============================================>

=over 4

=item B<DESCRIPTION>

 Portage's perl magicstream.  Tries to deduce if the input is compressed, piped
 or just a normal file and does the proper thing to open the stream.

=item B<SYNOPSIS>

 portage_utils::zin($STREAM, $stream_name)
 portage_utils::zin(*STREAM, $stream_name)
 portage_utils::zin(*STREAM, "plain-text-input-file");
 portage_utils::zin(*STREAM, "< plain-text-input-file");
 portage_utils::zin(*STREAM, "compressed-input-file.gz");
 portage_utils::zin(*STREAM, "< compressed-input-file.gz");
 portage_utils::zin(*STREAM, "input pipe |");

 @PARAM $STREAM
 @PARAM $stream_name

 @RETURN Returns 0 if the stream was opened.

=item B<SEE ALSO>

 zout()

=back

=cut

# ----------------------------------------------------------------------------#
sub zin($$) {
   # if no argument is sent in
   if(not @_) {
      die "\n!!! ERROR: portage_utils::zin requires input arguments! \taborting...\n";
   }
   # if we are called as a module then skip the our reference name
   elsif($_[0] and $_[0] =~ /^portage_utils/) {
      shift;
      # if no argument is sent in
      if(not @_) {
         die "\n!!! ERROR: portage_utils::zin requires input arguments! \taborting...\n";
      }
   }

   my($STREAM, $stream_name) = @_;

   print "DEBUG zin with $stream_name.\n" if ($DEBUG);

   die "Don't use zin to open an output stream!!\n" if ($stream_name =~ /^\s*>/);

   # remove leading < if present
   $stream_name =~ s/^\s*<//;

   # Is this a pipe.
   if ($stream_name =~ /\|\s*$/) {
      print "DEBUG zin is a pipe.\n" if ($DEBUG);
      return open($STREAM, "$stream_name");
   }
   # Is this a bzipped 2 stream.
   elsif ($stream_name =~ /$isBzip2/) {
      print "DEBUG zin is bzip2.\n" if ($DEBUG);
      return open($STREAM, "bzip2 -cqfd $stream_name |");
   }
   # Is this a gzipped stream.
   elsif ($stream_name =~ /$isZip/) {
      print "DEBUG zin is zip.\n" if (defined $DEBUG);
      return open($STREAM, "gzip -cqfd $stream_name |");
   }
   # Is this a lzma stream.
   elsif ($stream_name =~ /$isLzma/) {
      print "DEBUG zin is lzma.\n" if ($DEBUG);
      return open($STREAM, "lzma -4 -cqfd $stream_name |");
   }
   # Assume it is a normal file.
   else {
      print "DEBUG zin is plain.\n" if ($DEBUG);
      return open($STREAM, "<$stream_name");
   }
}



# ================= zout ====================#

=head1 SUB

B< =============================================
 ====== zout                             ======
 =============================================>

=over 4

=item B<DESCRIPTION>

 Portage's perl magicstream.  Tries to deduce if the output is compressed, piped
 or just a normal file and does the proper thing to open the stream.

=item B<SYNOPSIS>

 portage_utils::zout($STREAM, $stream_name)
 portage_utils::zout(*STREAM, $stream_name)
 portage_utils::zout(*STREAM, "> plain-text-output-file");
 portage_utils::zout(*STREAM, "> compressed-output-file.gz");
 portage_utils::zout(*STREAM, "| output pipe ");

 @PARAM $STREAM
 @PARAM $stream_name

 @RETURN Returns 0 if the stream was opened.

=item B<SEE ALSO>

 zin()

=back

=cut

# ----------------------------------------------------------------------------#
sub zout($$) {
   # if no argument is sent in
   if(not @_) {
      die "\n!!! ERROR: portage_utils::zout requires input arguments! \taborting...\n";
   }
   # if we are called as a module then skip the our reference name
   elsif($_[0] and $_[0] =~ /^portage_utils/) {
      shift;
      # if no argument is sent in
      if(not @_) {
         die "\n!!! ERROR: portage_utils::zout requires input arguments! \taborting...\n";
      }
   }

   my($STREAM, $stream_name) = @_;

   print "DEBUG zout with $stream_name.\n" if ($DEBUG);

   die "Don't use zout to open an input stream!!\n" if ($stream_name =~ /^\s*</);

   # Remove one redir.
   $stream_name =~ s/^\s*>//;

   # Is this a pipe.
   if ($stream_name =~ /^\s*\|/) {
      print "DEBUG zout is a pipe.\n" if ($DEBUG);
      die "Can't append to a pipe." if ($stream_name =~ /^\s*>/);
      return open($STREAM, "$stream_name");
   }
   # Is this a bzipped stream.
   elsif ($stream_name =~ /$isBzip2/) {
      print "DEBUG zout is bzip2.\n" if ($DEBUG);
      return open($STREAM, "| bzip2 -cqf >$stream_name");
   }
   # Is this a gzipped stream.
   elsif ($stream_name =~ /$isZip/) {
      print "DEBUG zout is zip.\n" if ($DEBUG);
      return open($STREAM, "| gzip -cqf >$stream_name");
   }
   # Is this a lzma stream.
   elsif ($stream_name =~ /$isLzma/) {
      print "DEBUG zout is lzma.\n" if ($DEBUG);
      die "Can't append to a lzma file." if ($stream_name =~ /^\s*>/);
      return open($STREAM, "| lzma -cqf >$stream_name");
   }
   # Assume it is a normal file.
   else {
      print "DEBUG zout is plain.\n" if ($DEBUG);
      $stream_name =~ s/^\s*>\s*-\s*$/-/;
      return open($STREAM, ">$stream_name");
   }
}

# The library was properly loaded.
1;
