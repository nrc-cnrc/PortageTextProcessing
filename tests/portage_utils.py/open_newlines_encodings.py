#!/usr/bin/env python3

import builtins
import os
import portage_utils
import time
from portage_utils import open

with builtins.open("newlines.txt", "wt", newline="", encoding="utf8") as f:
    f.write("TestéàûLF\nCR\rLF\nCRLF\r\nLF\n")
os.system("gzip < newlines.txt > newlines.gz")


for infile in "newlines.txt", "newlines.gz", "cat newlines.txt |":
    for encoding in None, "latin1", "utf8", portage_utils.DEFAULT_ENCODING_VALUE:
        for newline in None, "", "\n", "\r", "\r\n":
            print(
                infile,
                "default" if encoding is portage_utils.DEFAULT_ENCODING_VALUE else encoding,
                repr(newline),
                list(open(infile, "r", encoding=encoding, newline=newline))
            )
    print(
        infile,
        "rb",
        list(open(infile, "rb"))
    )


for outfile, result in (
    ("outlines.txt", "outlines.txt"),
    ("outlines.gz", "outlines.gz"),
    ("| cat > outlines.txt", "outlines.txt")
):
    for encoding in "latin1", "utf8", portage_utils.DEFAULT_ENCODING_VALUE:
        for newline in None, "", "\n", "\r", "\r\n":
            with open(outfile, "w", encoding=encoding, newline=newline) as f:
                print("àéCR\rCRLF\r", file=f)
                print("LF", file=f)
            time.sleep(0.1)
            with open(result, "rb") as res:
                print(
                    outfile,
                    "default" if encoding is portage_utils.DEFAULT_ENCODING_VALUE else encoding,
                    repr(newline),
                    repr(res.read()),
                )
    with open(outfile, "wb") as f:
        f.write(b"CR\rCRLF\r\nLF\n")
    time.sleep(0.1)
    with open(result, "rb") as res:
        print(
            outfile,
            "wb",
            repr(res.read()),
        )


