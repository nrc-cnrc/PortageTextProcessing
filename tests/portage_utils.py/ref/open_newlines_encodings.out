newlines.txt None None ['TestéàûLF\n', 'CR\n', 'LF\n', 'CRLF\n', 'LF\n']
newlines.txt None '' ['TestéàûLF\n', 'CR\r', 'LF\n', 'CRLF\r\n', 'LF\n']
newlines.txt None '\n' ['TestéàûLF\n', 'CR\rLF\n', 'CRLF\r\n', 'LF\n']
newlines.txt None '\r' ['TestéàûLF\nCR\r', 'LF\nCRLF\r', '\nLF\n']
newlines.txt None '\r\n' ['TestéàûLF\nCR\rLF\nCRLF\r\n', 'LF\n']
newlines.txt latin1 None ['TestÃ©Ã\xa0Ã»LF\n', 'CR\n', 'LF\n', 'CRLF\n', 'LF\n']
newlines.txt latin1 '' ['TestÃ©Ã\xa0Ã»LF\n', 'CR\r', 'LF\n', 'CRLF\r\n', 'LF\n']
newlines.txt latin1 '\n' ['TestÃ©Ã\xa0Ã»LF\n', 'CR\rLF\n', 'CRLF\r\n', 'LF\n']
newlines.txt latin1 '\r' ['TestÃ©Ã\xa0Ã»LF\nCR\r', 'LF\nCRLF\r', '\nLF\n']
newlines.txt latin1 '\r\n' ['TestÃ©Ã\xa0Ã»LF\nCR\rLF\nCRLF\r\n', 'LF\n']
newlines.txt utf8 None ['TestéàûLF\n', 'CR\n', 'LF\n', 'CRLF\n', 'LF\n']
newlines.txt utf8 '' ['TestéàûLF\n', 'CR\r', 'LF\n', 'CRLF\r\n', 'LF\n']
newlines.txt utf8 '\n' ['TestéàûLF\n', 'CR\rLF\n', 'CRLF\r\n', 'LF\n']
newlines.txt utf8 '\r' ['TestéàûLF\nCR\r', 'LF\nCRLF\r', '\nLF\n']
newlines.txt utf8 '\r\n' ['TestéàûLF\nCR\rLF\nCRLF\r\n', 'LF\n']
newlines.txt default None ['TestéàûLF\n', 'CR\n', 'LF\n', 'CRLF\n', 'LF\n']
newlines.txt default '' ['TestéàûLF\n', 'CR\r', 'LF\n', 'CRLF\r\n', 'LF\n']
newlines.txt default '\n' ['TestéàûLF\n', 'CR\rLF\n', 'CRLF\r\n', 'LF\n']
newlines.txt default '\r' ['TestéàûLF\nCR\r', 'LF\nCRLF\r', '\nLF\n']
newlines.txt default '\r\n' ['TestéàûLF\nCR\rLF\nCRLF\r\n', 'LF\n']
newlines.txt rb [b'Test\xc3\xa9\xc3\xa0\xc3\xbbLF\n', b'CR\rLF\n', b'CRLF\r\n', b'LF\n']
newlines.gz None None ['TestéàûLF\n', 'CR\n', 'LF\n', 'CRLF\n', 'LF\n']
newlines.gz None '' ['TestéàûLF\n', 'CR\r', 'LF\n', 'CRLF\r\n', 'LF\n']
newlines.gz None '\n' ['TestéàûLF\n', 'CR\rLF\n', 'CRLF\r\n', 'LF\n']
newlines.gz None '\r' ['TestéàûLF\nCR\r', 'LF\nCRLF\r', '\nLF\n']
newlines.gz None '\r\n' ['TestéàûLF\nCR\rLF\nCRLF\r\n', 'LF\n']
newlines.gz latin1 None ['TestÃ©Ã\xa0Ã»LF\n', 'CR\n', 'LF\n', 'CRLF\n', 'LF\n']
newlines.gz latin1 '' ['TestÃ©Ã\xa0Ã»LF\n', 'CR\r', 'LF\n', 'CRLF\r\n', 'LF\n']
newlines.gz latin1 '\n' ['TestÃ©Ã\xa0Ã»LF\n', 'CR\rLF\n', 'CRLF\r\n', 'LF\n']
newlines.gz latin1 '\r' ['TestÃ©Ã\xa0Ã»LF\nCR\r', 'LF\nCRLF\r', '\nLF\n']
newlines.gz latin1 '\r\n' ['TestÃ©Ã\xa0Ã»LF\nCR\rLF\nCRLF\r\n', 'LF\n']
newlines.gz utf8 None ['TestéàûLF\n', 'CR\n', 'LF\n', 'CRLF\n', 'LF\n']
newlines.gz utf8 '' ['TestéàûLF\n', 'CR\r', 'LF\n', 'CRLF\r\n', 'LF\n']
newlines.gz utf8 '\n' ['TestéàûLF\n', 'CR\rLF\n', 'CRLF\r\n', 'LF\n']
newlines.gz utf8 '\r' ['TestéàûLF\nCR\r', 'LF\nCRLF\r', '\nLF\n']
newlines.gz utf8 '\r\n' ['TestéàûLF\nCR\rLF\nCRLF\r\n', 'LF\n']
newlines.gz default None ['TestéàûLF\n', 'CR\n', 'LF\n', 'CRLF\n', 'LF\n']
newlines.gz default '' ['TestéàûLF\n', 'CR\r', 'LF\n', 'CRLF\r\n', 'LF\n']
newlines.gz default '\n' ['TestéàûLF\n', 'CR\rLF\n', 'CRLF\r\n', 'LF\n']
newlines.gz default '\r' ['TestéàûLF\nCR\r', 'LF\nCRLF\r', '\nLF\n']
newlines.gz default '\r\n' ['TestéàûLF\nCR\rLF\nCRLF\r\n', 'LF\n']
newlines.gz rb [b'Test\xc3\xa9\xc3\xa0\xc3\xbbLF\n', b'CR\rLF\n', b'CRLF\r\n', b'LF\n']
cat newlines.txt | None None ['TestéàûLF\n', 'CR\n', 'LF\n', 'CRLF\n', 'LF\n']
cat newlines.txt | None '' ['TestéàûLF\n', 'CR\r', 'LF\n', 'CRLF\r\n', 'LF\n']
cat newlines.txt | None '\n' ['TestéàûLF\n', 'CR\rLF\n', 'CRLF\r\n', 'LF\n']
cat newlines.txt | None '\r' ['TestéàûLF\nCR\r', 'LF\nCRLF\r', '\nLF\n']
cat newlines.txt | None '\r\n' ['TestéàûLF\nCR\rLF\nCRLF\r\n', 'LF\n']
cat newlines.txt | latin1 None ['TestÃ©Ã\xa0Ã»LF\n', 'CR\n', 'LF\n', 'CRLF\n', 'LF\n']
cat newlines.txt | latin1 '' ['TestÃ©Ã\xa0Ã»LF\n', 'CR\r', 'LF\n', 'CRLF\r\n', 'LF\n']
cat newlines.txt | latin1 '\n' ['TestÃ©Ã\xa0Ã»LF\n', 'CR\rLF\n', 'CRLF\r\n', 'LF\n']
cat newlines.txt | latin1 '\r' ['TestÃ©Ã\xa0Ã»LF\nCR\r', 'LF\nCRLF\r', '\nLF\n']
cat newlines.txt | latin1 '\r\n' ['TestÃ©Ã\xa0Ã»LF\nCR\rLF\nCRLF\r\n', 'LF\n']
cat newlines.txt | utf8 None ['TestéàûLF\n', 'CR\n', 'LF\n', 'CRLF\n', 'LF\n']
cat newlines.txt | utf8 '' ['TestéàûLF\n', 'CR\r', 'LF\n', 'CRLF\r\n', 'LF\n']
cat newlines.txt | utf8 '\n' ['TestéàûLF\n', 'CR\rLF\n', 'CRLF\r\n', 'LF\n']
cat newlines.txt | utf8 '\r' ['TestéàûLF\nCR\r', 'LF\nCRLF\r', '\nLF\n']
cat newlines.txt | utf8 '\r\n' ['TestéàûLF\nCR\rLF\nCRLF\r\n', 'LF\n']
cat newlines.txt | default None ['TestéàûLF\n', 'CR\n', 'LF\n', 'CRLF\n', 'LF\n']
cat newlines.txt | default '' ['TestéàûLF\n', 'CR\r', 'LF\n', 'CRLF\r\n', 'LF\n']
cat newlines.txt | default '\n' ['TestéàûLF\n', 'CR\rLF\n', 'CRLF\r\n', 'LF\n']
cat newlines.txt | default '\r' ['TestéàûLF\nCR\r', 'LF\nCRLF\r', '\nLF\n']
cat newlines.txt | default '\r\n' ['TestéàûLF\nCR\rLF\nCRLF\r\n', 'LF\n']
cat newlines.txt | rb [b'Test\xc3\xa9\xc3\xa0\xc3\xbbLF\n', b'CR\rLF\n', b'CRLF\r\n', b'LF\n']
outlines.txt latin1 None b'\xe0\xe9CR\rCRLF\r\nLF\n'
outlines.txt latin1 '' b'\xe0\xe9CR\rCRLF\r\nLF\n'
outlines.txt latin1 '\n' b'\xe0\xe9CR\rCRLF\r\nLF\n'
outlines.txt latin1 '\r' b'\xe0\xe9CR\rCRLF\r\rLF\r'
outlines.txt latin1 '\r\n' b'\xe0\xe9CR\rCRLF\r\r\nLF\r\n'
outlines.txt utf8 None b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
outlines.txt utf8 '' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
outlines.txt utf8 '\n' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
outlines.txt utf8 '\r' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\rLF\r'
outlines.txt utf8 '\r\n' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\r\nLF\r\n'
outlines.txt default None b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
outlines.txt default '' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
outlines.txt default '\n' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
outlines.txt default '\r' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\rLF\r'
outlines.txt default '\r\n' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\r\nLF\r\n'
outlines.txt wb b'CR\rCRLF\r\nLF\n'
outlines.gz latin1 None b'\xe0\xe9CR\rCRLF\r\nLF\n'
outlines.gz latin1 '' b'\xe0\xe9CR\rCRLF\r\nLF\n'
outlines.gz latin1 '\n' b'\xe0\xe9CR\rCRLF\r\nLF\n'
outlines.gz latin1 '\r' b'\xe0\xe9CR\rCRLF\r\rLF\r'
outlines.gz latin1 '\r\n' b'\xe0\xe9CR\rCRLF\r\r\nLF\r\n'
outlines.gz utf8 None b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
outlines.gz utf8 '' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
outlines.gz utf8 '\n' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
outlines.gz utf8 '\r' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\rLF\r'
outlines.gz utf8 '\r\n' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\r\nLF\r\n'
outlines.gz default None b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
outlines.gz default '' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
outlines.gz default '\n' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
outlines.gz default '\r' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\rLF\r'
outlines.gz default '\r\n' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\r\nLF\r\n'
outlines.gz wb b'CR\rCRLF\r\nLF\n'
| cat > outlines.txt latin1 None b'\xe0\xe9CR\rCRLF\r\nLF\n'
| cat > outlines.txt latin1 '' b'\xe0\xe9CR\rCRLF\r\nLF\n'
| cat > outlines.txt latin1 '\n' b'\xe0\xe9CR\rCRLF\r\nLF\n'
| cat > outlines.txt latin1 '\r' b'\xe0\xe9CR\rCRLF\r\rLF\r'
| cat > outlines.txt latin1 '\r\n' b'\xe0\xe9CR\rCRLF\r\r\nLF\r\n'
| cat > outlines.txt utf8 None b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
| cat > outlines.txt utf8 '' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
| cat > outlines.txt utf8 '\n' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
| cat > outlines.txt utf8 '\r' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\rLF\r'
| cat > outlines.txt utf8 '\r\n' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\r\nLF\r\n'
| cat > outlines.txt default None b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
| cat > outlines.txt default '' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
| cat > outlines.txt default '\n' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\nLF\n'
| cat > outlines.txt default '\r' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\rLF\r'
| cat > outlines.txt default '\r\n' b'\xc3\xa0\xc3\xa9CR\rCRLF\r\r\nLF\r\n'
| cat > outlines.txt wb b'CR\rCRLF\r\nLF\n'
